core/lib/tasks/payment_method.rake

```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+
+namespace :payment_method do
+  desc "Deprecates old payment methods and fixes ActiveRecord::SubclassNotFound error, "\
+  "which happens after switching Payment Service Provider."
+  task deprecate_unsupported_payment_methods: :environment do
+    unsupported_ids = get_unsupported_payment_method_ids
```

Comment:
Maybe I'm missing something but what about removing these methods and just leave the code to execute the change when the exception is raised? Something like:

```ruby
    Spree::PaymentMethod.pluck(:type).select do |type|
      begin
        ActiveSupport::Dependencies.constantize(type)
      rescue NameError
        # do the updates here
      end
    end
```

Reply:
I'm a bit confused by what you quoted, I'll try to answer for what I understood.
I personally wouldn't apply the fix automatically when the STI error happens: that error reports an inconsistent situation on the db and fixing it actually changes a bit of the behavior of the application (querying for payment methods will return instances of another class after that).
I don't like the idea of this happening without the store devs understanding this problem, and I thought that they may need to take more actions than just this when this problem occurs.

But I lack of experience on this project, so please correct me if my assumption is wrong.

About the code you inserted: that rescue instruction is where the task understands what types are not supported anymore, and not where the original error is caught to raise a more understandable error.
The error is caught in `Spree::PaymentMethod#find_sti_class,` so it should be caught there in order to make this automatic. Did I understand correctly what you meant?

Reply:
Sorry, my comment wasn't clear at all: I meant to do something like this directly in the rake task:

```ruby
  task deprecate_unsupported_payment_methods: :environment do
    Spree::PaymentMethod.pluck(:id, :type).each do |id, type|
      begin
        ActiveSupport::Dependencies.constantize(type)
      rescue NameError
        # we have the id of the problematic type here, can we avoid the extra query and just update the record?
      end
    end
  end
```

Reply:
It's possible, the records will be a few, so I can see no performance issue. It would make much more sense merging the two update statements in the first one, as you asked above. I'll wait for you to answer to that before changing this.